exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java

```diff
@@ -169,50 +211,126 @@ protected void buildSchema() throws SchemaChangeException {
+          if ( inputIndex == 0 ) {
+            // Indicate that a schema was seen (in case probe side is empty)
+            probeSchema = probeBatch.getSchema();
+          } else {
+            // We need to have the schema of the build side even when the build side is empty
+            rightSchema = buildBatch.getSchema();
+            // position of the new "column" for keeping the hash values (after the real columns)
+            rightHVColPosition = buildBatch.getContainer().getNumberOfColumns();
+            // new schema can also contain records
+          }
+        case OK:
+          if (recordBatch.getRecordCount() == 0) {
+            continue;
+          }
+          // We got a non empty batch
+        default:
+          // Other cases termination conditions
+          return outcome;
+      }
+    }
+  }
+
+  /**
+   * Determines the memory calculator to use. If maxNumBatches is configured simple batch counting is used to spill. Otherwise
+   * memory calculations are used to determine when to spill.
+   * @return The memory calculator to use.
+   */
+  public HashJoinMemoryCalculator getCalculatorImpl() {
+    if (maxBatchesInMemory == 0) {
+      final double safetyFactor = context.getOptions().getDouble(ExecConstants.HASHJOIN_SAFETY_FACTOR_KEY);
+      final double fragmentationFactor = context.getOptions().getDouble(ExecConstants.HASHJOIN_FRAGMENTATION_FACTOR_KEY);
+      final double hashTableDoublingFactor = context.getOptions().getDouble(ExecConstants.HASHJOIN_HASH_DOUBLE_FACTOR_KEY);
+      final String hashTableCalculatorType = context.getOptions().getString(ExecConstants.HASHJOIN_HASHTABLE_CALC_TYPE_KEY);
+
+      return new HashJoinMemoryCalculatorImpl(safetyFactor, fragmentationFactor, hashTableDoublingFactor, hashTableCalculatorType);
+    } else {
+      return new HashJoinMechanicalMemoryCalculator(maxBatchesInMemory);
+    }
   }
 
   @Override
   public IterOutcome innerNext() {
+    // In case incoming was killed before, just cleanup and return
+    if ( wasKilled ) {
+      this.cleanup();
+      super.close();
+      return IterOutcome.NONE;
+    }
+
     try {
       /* If we are here for the first time, execute the build phase of the
        * hash join and setup the run time generated class for the probe side
        */
       if (state == BatchState.FIRST) {
         // Build the hash table, using the build side record batches.
         executeBuildPhase();
-        hashJoinProbe.setupHashJoinProbe(context, hyperContainer, left, left.getRecordCount(), this, hashTable,
-            hjHelper, joinType, leftUpstream);
         // Update the hash table related stats for the operator
-        updateStats(this.hashTable);
+        updateStats();
+        // Initialize various settings for the probe side
+        hashJoinProbe.setupHashJoinProbe(probeBatch, this, joinType, leftUpstream, partitions, cycleNum, container, spilledInners, buildSideIsEmpty, numPartitions, rightHVColPosition);
```

Comment:
see other comments about line too long.

Reply:
LGTM.  +1   @Ben-Zvi thanks for adding the HashJoinProbeTemplate.  Pls see couple of cosmetic comments.